File: nodejs/src/client/activity.ts
/**
 * Activity API: list activity and audit entries (unified read-model query).
 * Canonical API: GET /ai/activity.
 * Synchronous read (R18.6 carve-out).
 *
 * Requirements: 7.4, 18.5
 */

import type { LoxtepHttpClient } from '../http/client.js';
import type {
  ActivityEntry,
  ActivityListFilters,
  ActivityListResponse,
} from './activity-types.js';

const ACTIVITY_BASE = '/ai/activity';

function buildQueryString(params: Record<string, string | number | undefined>): string {
  const search = new URLSearchParams();
  for (const [k, v] of Object.entries(params)) {
    if (v === undefined || v === null) continue;
    search.set(k, String(v));
  }
  const qs = search.toString();
  return qs ? `?${qs}` : '';
}

export interface ActivityApi {
  /**
   * List activity and audit entries ordered by UTC timestamp DESC with stable entry_id tie-break.
   * Applies source/actor/resource-type/time-range filters combined with AND (R7.4).
   * Synchronous read — R18.6 carve-out.
   */
  list: (filters?: ActivityListFilters) => Promise<{ entries: ActivityEntry[]; cursor: string | null }>;
}

export function createActivityApi(http: LoxtepHttpClient): ActivityApi {
  return {
    async list(filters?: ActivityListFilters) {
      const params: Record<string, string | number | undefined> = {};
      if (filters?.source) params.source = filters.source;
      if (filters?.actor) params.actor = filters.actor;
      if (filters?.resource_type) params.resource_type = filters.resource_type;
      if (filters?.start) params.start = filters.start;
      if (filters?.end) params.end = filters.end;
      if (filters?.limit != null) params.limit = filters.limit;
      if (filters?.cursor) params.cursor = filters.cursor;
      const qs = buildQueryString(params);
      const res = await http.get<ActivityListResponse>(`${ACTIVITY_BASE}${qs}`);
      return res.data;
    },
  };
}

File: python/src/loxtep/activity.py
"""
Activity API (activity / audit entries). list.
Internal / experimental — not part of the documented customer surface.
Backend: /ai/activity.
"""

from typing import Any, Optional

from .http_client import AsyncLoxtepHttpClient, LoxtepHttpClient

BASE = "/ai/activity"


def _unwrap(res: Any) -> Any:
    return res.get("data", res) if isinstance(res, dict) else res


def _query_string(params: dict[str, Any]) -> str:
    parts = [f"{k}={v}" for k, v in params.items() if v is not None]
    return "?" + "&".join(parts) if parts else ""


class ActivityApi:
    """Sync activity surface (internal)."""

    def __init__(self, http: LoxtepHttpClient) -> None:
        self._http = http

    def list(
        self,
        *,
        source: Optional[str] = None,
        actor: Optional[str] = None,
        resource_type: Optional[str] = None,
        start: Optional[str] = None,
        end: Optional[str] = None,
        limit: Optional[int] = None,
        cursor: Optional[str] = None,
    ) -> dict[str, Any]:
        qs = _query_string(
            {
                "source": source,
                "actor": actor,
                "resource_type": resource_type,
                "start": start,
                "end": end,
                "limit": limit,
                "cursor": cursor,
            }
        )
        return _unwrap(self._http.get(f"{BASE}{qs}"))


class AsyncActivityApi:
    """Async activity surface (internal)."""

    def __init__(self, http: AsyncLoxtepHttpClient) -> None:
        self._http = http

    async def list(
        self,
        *,
        source: Optional[str] = None,
        actor: Optional[str] = None,
        resource_type: Optional[str] = None,
        start: Optional[str] = None,
        end: Optional[str] = None,
        limit: Optional[int] = None,
        cursor: Optional[str] = None,
    ) -> dict[str, Any]:
        qs = _query_string(
            {
                "source": source,
                "actor": actor,
                "resource_type": resource_type,
                "start": start,
                "end": end,
                "limit": limit,
                "cursor": cursor,
            }
        )
        return _unwrap(await self._http.get(f"{BASE}{qs}"))
